chainer/functions/loss/discriminative_loss.py

```diff
@@ -0,0 +1,378 @@
+import chainer
+from chainer.backends import cuda
+from chainer.functions.array.broadcast import broadcast_to
+from chainer.functions.array.cast import cast
+from chainer.functions.array.concat import concat
+from chainer.functions.array.expand_dims import expand_dims
+from chainer.functions.array.reshape import reshape
+from chainer.functions.array.stack import stack
+from chainer.functions.array.transpose import transpose
+from chainer.functions.math.average import average
+from chainer.functions.math.basic_math import absolute
+from chainer.functions.math.maximum import maximum
+from chainer.functions.math.sum import sum
+
+
+class DiscriminativeMarginBasedClusteringLoss(object):
+    """Discriminative margin based clustering loss function
+
+    This is the implementation of following paper:
+    https://arxiv.org/pdf/1802.05591.pdf
+
+    In segmentation, one of the biggest problem is having noise at the output
+    of trained network.
+    For cross-entropy based approaches, if the pixel value is wrong,
+    the loss value will be same independent from wrong pixel's location.
+    However, for segmentation, even though network gives wrong pixel output,
+    it is desirable to have it
+    as close as possible to the original position.
+    By applying discriminative loss function, groups of segmentation instances
+    can be moved together.
+
+    This loss function calculates three different parameters:
+        - Variance Loss:
+            Loss to penalize distances between pixels which are belonging
+            to same instance. (Pull force)
+        - Distance loss:
+            Loss to penalize distances between the centers of instances.
+            (Push force)
+        - Regularization loss:
+            Small regularization loss to penalize weights against overfit
+
+    Args:
+        delta_v (float): Minimum distance to start penalizing variance
+        delta_d (float): Maximum distance to stop penalizing distance
+        max_n_clusters (int): Maximum possible number of clusters.
+        norm (int): Norm to calculate pixels and cluster center distances
+        alpha (float): Weight for variance loss      (alpha * variance_loss)
+        beta (float): Weight for distance loss       (beta * distance_loss)
+        gamma (float): Weight for regularization loss(gamma * regularizer_loss)
+
+    Returns:
+        float: (alpha * variance_loss)+
```

Comment:
I think the returned value is not a float but a Chainer variable holding a NumPy or CuPy array whose shape is `()`.

Reply:
I found sphinx-build raises a syntax error at this line. Could you investigate and fix it?

```
sphinx-build -b html -d build/doctrees source build/html

...(omit)...

Exception occurred:
  File "/Users/oonokenta/dev/chainer/docs/source/_docstring_check.py", line 46, in error
    raise InvalidDocstringError(msg, self, iline)
_docstring_check.InvalidDocstringError: Invalid indentation of `Returns` section

on chainer.functions.discriminative_margin_based_clustering_loss
at line 55: "                 (beta * distance_loss) +"

docstring:
00 Discriminative margin based clustering loss function
01 
02 This is the implementation of following paper:
03 https://arxiv.org/pdf/1802.05591.pdf
04 
05 In segmentation, one of the biggest problem is having noise at the output
06 of a trained network.
07 For cross-entropy based approaches, if the pixel value is wrong,
08 the loss value will be same independent from wrong pixel's location.
09 However, for segmentation, even though network gives wrong pixel output,
10 it is desirable to have it
11 as close as possible to the original position.
12 By applying discriminative loss function, groups of segmentation instances
13 can be moved together.
14 
15 This loss function calculates three different parameters:
16     - Variance Loss:
17         Loss to penalize distances between pixels which are belonging
18         to same instance. (Pull force)
19     - Distance loss:
20         Loss to penalize distances between the centers of instances.
21         (Push force)
22     - Regularization loss:
23         Small regularization loss to penalize weights against overfit
24 
25 :param x: Contains several inputs
26           - x[0] = segmentation prediction output     (n, i, w, h)
27           - x[1] = segmentation ground truth          (n, i, w, h)
28           - x[2] = number of objects in ground truth  (n,)
29           - x[3] = indexes of non-zero ground truths  (n, variable)
30 :type x: tuple
31 :param where,:
32                - n is batch size
33                - i is total instance count
34                - w is width of the image
35                - h is height of the image
36                - variable is variable length depending on list size
37 :param delta_v: Minimum distance to start penalizing variance
38 :type delta_v: float
39 :param delta_d: Maximum distance to stop penalizing distance
40 :type delta_d: float
41 :param max_n_clusters: Maximum possible number of clusters.
42 :type max_n_clusters: int
43 :param norm: Norm to calculate pixels and cluster center distances
44 :type norm: int
45 :param alpha: Weight for variance loss      (alpha * variance_loss)
46 :type alpha: float
47 :param beta: Weight for distance loss       (beta * distance_loss)
48 :type beta: float
49 :param gamma: Weight for regularization loss(gamma * regularizer_loss)
50 :type gamma: float
51 
52 :returns:
53 
54           (alpha * variance_loss)+
55                  (beta * distance_loss) +
56                  (gamma * regularizer_loss)
57 :rtype: float
58
```